sklearn/metrics/_scorer.py

```diff
@@ -771,6 +771,42 @@ def __getitem__(self, item):
         return super().__getitem__(item)
 
 
+def oob_score(estimator, X, y):
+    """Out-of-Bag scorer.
+
+    This function can be used wherever a scoring function is expected, but delegates
+    its functionality to the oob_score_ field of certain classifiers if they have one,
+    notably the RandomForest set of Estimators.
+
+    Parameters
+    ----------
+    estimator : BaseEstimator with an oob_score_ field
+        Estimator to assess.
+    X : array-like of shape (n_samples, n_variables)
+        Test set data.
+    y : array-like of shape (n_samples,)
+        Ground truth for the test set.
+
+    Returns
+    -------
+    OOB : float
+        Out-of-bag score.
+    """
+    if hasattr(estimator, "oob_score_"):
+        return estimator.oob_score_
```

Comment:
If a `RandomForestClassifer` is in a `Pipeline`, then getting the attribute will be more complicated.

We can special case for `Pipeline`, but that will not work generically. There can be third party meta-estimators that can wrap a `RandomForestClassifer` that we can not special case for.

Reply:
Is there no "get final estimator" method or function that is guaranteed to return the actual trained model from a pipeline that we can use? Ideally the special case won't have to live in the scorer but in a more re-usable place.

Reply:
One can special case for a `Pipeline` by doing `pipe[-1].oob_score_`. But that does not take into account all the different ways a third party estimator or meta-estimator may wrap a `RandomForest`.

Not being generic enough suggest that this function is better as an example than a function in the library. A user can already provide their own custom `oob_score` callable that works with their custom estimator.

Reply:
I would strongly prefer this doesn't become an example, it will see much less use if it is.

If the meta estimators can't access estimator attributes, how do we access the `feature_importance_` when using something like RFECV? Surely there is a mechanism for this. Also I note that there is positive response to #23400 which might mean this is generalizable to several estimator classes.

Reply:
Also, if you don't like it as a global class, how about as a mixin for the forest-based classifiers, such that you can
```python
cv = GridSearchCV(
    RandomForestClassifier(oob_score=True, random_state=0),
    {"n_estimators": [1, 20, 100]},
    cv=RandomForestClassifier.oob_cv,
    scoring=RandomForestClassifier.oob_scorer,
)
results = cv.fit(X, y)
```

Reply:
That could work for `cv`, but it is not standard practice for `sklearn`. The scorer would still have issues. The idea is to work with pipelines:

```python
clf = Pipeline(
    steps=[("preprocessor", preprocessor), ("classifier", RandomForestClassifier(oob_score=True))]
)

cv = GridSearchCV(
    clf,
    {"classifier__max_features": ["sqrt", "log2"]},
    cv=RandomForestClassifier.oob_cv,
    scoring=lambda est, X, y: est[-1].oob_score_
)
results = cv.fit(X, y)
```

Using `RandomForestClassifier.oob_scorer` does not work because it does not know it is in a Pipeline.

What if someone has a meta-estimator at the end, lets say a `VotingClassifier` and bags them?

```python
voting = VotingClassifier([
    BaggingClassifer(LogisticRegression(), oob_score=True),
    RandomForestClassifier(oob_score=True),
])
clf = Pipeline(
    steps=[("preprocessor", preprocessor), ("classifier", voting)]
)

def mean_oob_score_(est, X, y):
    # est is a pipeline with a voting classifer at the end.
    return np.mean([e.oob_score_ in est[-1].estimators_])

cv = GridSearchCV(
    clf,
    {"classifier__max_features": ["sqrt", "log2"]},
    cv=RandomForestClassifier.oob_cv,
    scoring=mean_oob_score_
)
results = cv.fit(X, y)
```

Reply:
> If the meta estimators can't access estimator attributes, how do we access the feature_importance_ when using something like RFECV?

[RFECV](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFECV.html#sklearn.feature_selection.RFECV) has a `importance_getter` allowing a user to define a custom way to get feature importances.

Reply:
Right so could I not just define an optional `score_getter` that defaults to `oob_score_`? I get that there are a few less estimators that have a built-in score than have feature importance, but it applies to all the forest classifiers I think. Plus XGBoost externally.

Reply:
There are two user stories I am concerned about:

### Going from RandomForest -> RandomForest in Pipeline

Let's say this PR is merged so users can do this:

```python
cv = GridSearchCV(
    RandomForestClassifier(...), ...
    cv=IdentitySplitter(), scoring=oob_score
)
```

If they switch to a pipeline because of categorical data, they would need to define their own custom `oob_score`. We can work around this by special casing pipeline, but I am not happy with special casing if there is an alternative.

### Unintended usage

Imagine a user only setting `scoring=oob_score` and not the `cv=IdentitySplitter` or the other way around.

## Possible way forward

Even with this PR, we would need an example to showcase the feature. Moving forward, I think we can have an example that defines a custom splitter and function and use them to show case oob_score for validation. As a follow up, we can decide on moving the Splitter and function into the library.

Reply:
Okay, fair enough. If you would like to see this as an example then I'll submit another PR which refactors this into one.

Reply:
Also, I suggest having this PR become the example PR. Maintainer time and reviewer time is limited, having one PRs helps focus attention.

Reply:
Understood. But I don't want to lose the work I've done, and I don't think I can switch branches etc.

Reply:
Since including the feature has not been decided yet, I think we can add an example on top of this PR with `IdentitySplitter` and `oob_score` being a part of the library.

If we decide to only have an example, then we can refactor this PR into a new one that only has the example.

Reply:
Resolving since none of this matters now that the classes aren't in the PR.